### ISSUE TITLE
Add `declaration-block-nesting-at-rule-required-list`

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#8481]
### What is the problem you're trying to solve?

I'd like to require `@layer` and other at-rules around my declarations.

Allow:

```css
@layer defaults {
  a { color: red; }
}
```

Disallow:

```css
a { color: red; }
```

### What solution would you like to see?

We typically enable users to do this on the [configuration level](https://stylelint.io/user-guide/customize) by adding generic and configurable rules rather than specific ones.

A new rule to do this. 

- Name: ~~`declaration-nesting-at-rule-required-list`~~ `declaration-block-nesting-at-rule-required-list`
- Description: Specify a list of required nesting at-rules for declarations.
- Message: `Expected nesting at-rule "${atRule}" for declaration`
- Primary option: `array[string|regex]|string|regex`
- Secondary options: None
- Category: Enforce conventions > [Allowed, disallowed & required](https://stylelint.io/user-guide/rules#allowed-disallowed--required)
- Autofix?: No
- Included in sharable configs?: No
- Expanded description: N/A

```json
{ 
  "rules": {
     "declaration-nesting-at-rule-required-list": ["layer", "scope"]
  }
}
```

Prior art: [use-layers](https://github.com/eslint/css/blob/main/docs/rules/use-layers.md).

**Comment by ybiquitous** [comment]
Sounds good, but let me confirm: is this rule for rules, not declarations? In the sample CSS code, `a { ... }` not in `@layer {}` seems to be disallowed.

**Comment by jeddy3** [comment]
Yes, it's ultimately for style rules, but it gets tricky with nesting. 

For example, this should be allowed:

```css
a {
  @layer {
     color: red;
  }
}
```

So I scoped it to declarations. The [spec](https://drafts.csswg.org/css-syntax-3/#block-contents) may give us a better name, but nothing stood out.

**Comment by ybiquitous** [comment]
Oh, the example of `a { @layer {...} }` is indeed tricky. I got the naming. 👍🏼 

So, what do you think about the reported ranges? For example, is the first declaration only be reported?

```css
a {
  @layer {
     color: red;
/*   ~~~~~~~~~~~ Expected nesting at-rule "${atRule}" for declaration */

     text-decoration: underline;
  }
}
```

**Comment by jeddy3** [comment]
Good point. I had initially thought to call it `declaration-list-nesting-at-rule-required-list`  where _declaration list_ is the thing being targeted, but didn't like the two `list`s. However, it's the more accurate name that maps better onto a reported range.

```css
a {
  @layer {
     /* start of range */ color: red;
     text-decoration: underline; /* end of range */ 
  }
}
```

Shall we go with `declaration-list-nesting-at-rule-required-list`?

**Comment by ybiquitous** [comment]
The reported range sounds good to me. 👍🏼 

The name `declaration-list-nesting-at-rule-required-list` is not so bad, but the doubled "list" seems a bit verbose to me. So, according to the [CSS spec](https://drafts.csswg.org/cssom/#css-declarations), does it make sense to use `declarations` or `declaration-block` instead of `declaration-list`? 🤔

E.g.

- `declarations-nesting-at-rule-required-list`
- `declaration-block-nesting-at-rule-required-list`

**Comment by jeddy3** [comment]
Block SGTM - we've used it in 5 previous rules, e.g. [duplicates](https://stylelint.io/user-guide/rules#duplicate).

I've labelled the issue ready to implement. If anyone has time, please consider [contributing](https://stylelint.io/contributing).

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

Possible references:

- [other `*-required-list` rule](https://github.com/stylelint/stylelint/tree/017aff1cbade861cd98c01c41f26096cd5d6dcd6/lib/rules/at-rule-property-required-list)
- [example of array as primary option](https://github.com/stylelint/stylelint/blob/017aff1cbade861cd98c01c41f26096cd5d6dcd6/lib/rules/at-rule-allowed-list/index.mjs#L21-L30)
- [example of `findNodeUpToRoot` util in use](https://github.com/stylelint/stylelint/blob/017aff1cbade861cd98c01c41f26096cd5d6dcd6/lib/rules/property-no-unknown/index.mjs#L95-L101)

**Comment by ybiquitous** [comment]
Sounds good. I just fixed the rule name in this issue description. 👍🏼

**Comment by sw1tch3roo** [comment]
I'll take this too If no one is against it 🐸

**Comment by jeddy3** [comment]
@sw1tch3roo That'd be fab.

As with https://github.com/stylelint/stylelint/pull/8674, let's start with a small range of tests that we can build on during the review process.

For example, given:

```
{ 
  "rules": {
     "declaration-nesting-at-rule-required-list": ["layer", "/^s/"]
  }
}
```

Accept:

```css
@layer {
  a { color: red; }
}
```

```css
a {
  @layer {
     color: red;
  }
}
```

```css
@supports (color: red) {
  a { color: red; }
}
```

```css
@scope (a) {
  a { color: red; }
}
```

```css
@media all {}
```

```css
@layer {
  @media all {
    a { color: red; }
  }
}
```

While disallowing:

```css
a { color: red; }
```

```css
a { 
  color: red;
  color: blue;
}
```

```css
a { color: red; }
.foo { color: red; }
```

```css
@media all {
  a { color: red; }
}
```

You can add more as you build the rule, though, and encounter paths and patterns you want to test.
